### PR TITLE
Add quarkus-bootstrap:build-tree

### PR DESCRIPTION
**Pull request by aloubyansky**
Fix for #1877, bootstrap maven plugin goal to display build dependency tree

### CODE REVIEW

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.maven.project.MavenProject;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.repository.RemoteRepository;
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.resolver.AppModelResolverException;
+import io.quarkus.bootstrap.resolver.BootstrapAppModelResolver;
+import io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver;
+
+@Mojo(name = "build-dependency-tree", defaultPhase = LifecyclePhase.NONE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
```
Comment (cescoffier) [inline independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java]: We could shorten the name no?

What about `build-tree` or even just `tree` as the Maven Dependency plugin?

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java [hunk 2]
```diff
@@ -41,12 +41,10 @@
...
 import org.junit.jupiter.api.extension.AfterEachCallback;
-import org.junit.jupiter.api.extension.BeforeAllCallback;
 import org.junit.jupiter.api.extension.BeforeEachCallback;
 import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.jupiter.api.extension.TestInstanceFactory;
 import org.junit.jupiter.api.extension.TestInstanceFactoryContext;
-import org.junit.jupiter.api.extension.TestInstancePostProcessor;
 import org.junit.jupiter.api.extension.TestInstantiationException;
 import org.junit.platform.commons.JUnitException;
 import org.objectweb.asm.ClassReader;
```
Comment (cescoffier) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java]: Could we have a test for this mojo just to be sure it "works"

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    public boolean visitEnter(DependencyNode node) {
+        StringBuilder buffer = new StringBuilder(128);
+        buffer.append(indent);
+        org.eclipse.aether.graph.Dependency dep = node.getDependency();
+        if (dep != null) {
+            org.eclipse.aether.artifact.Artifact art = dep.getArtifact();
+
+            buffer.append(art);
+            buffer.append(':').append(dep.getScope());
```
Comment (cescoffier) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java]: I can't remember if toString give the classifier and packaging / extension.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                        buf.append("   ");
+                    }
+                }
+                if (depth.get(depth.size() - 1)) {
+                    buf.append("|- ");
+                } else {
+                    buf.append("\\- ");
+                }
+            }
+            buf.append(dep.getArtifact()).append(':').append(dep.getScope());
```
Comment (cescoffier) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java]: Does it print the classifier and extension?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java]: Thanks, that's why I wrote to the dev list. I think there has to be a hint that these are the dependencies used for the build, not runtime or something. (potentially, we could add goals to displays other classpath trees, e.g. https://gist.githubusercontent.com/dmlloyd/c1aaadb9635e12b23581137c4ba9b841/raw/class-loaders.txt)

The maven analogue is specialized on dependencies, i.e. 'dependency:tree', while `quarkus-bootstrap` is dedicated to anything related to bootstrap.

I guess, `build-tree` could work though. How about `build-cp`?

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java (hunk 2 above)
Comment (aloubyansky) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java]: I'll have to figure out how to write one.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java (hunk 3 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java]: Thanks, this class is not used at all!

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java (hunk 4 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java]: Yes, it's the same format as in `dependency:tree`

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java (hunk 1 above)
Comment (cescoffier) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java]: `build-tree` looks good to me. Short and clear.

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java (hunk 2 above)
Comment (cescoffier) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java]: Look how the tests of the main Maven plugin are written. Basically, you provide a project and run it. Then, you check that nothing unexpected happened.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java (hunk 4 above)
Comment (cescoffier) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java]: Good to know!

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java (hunk 3 above)
Comment (cescoffier) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java]: Should it be deleted?

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java (hunk 2 above)
Comment (aloubyansky) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java]: Added a simple test.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java (hunk 3 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DependencyGraphLogger.java]: Sure, deleted.

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/BuildDependencyTreeMojo.java]: Agreed

**Comment by aloubyansky** [comment]
Hm, my unicode character drawing test didn't work on windows.

**Comment by aloubyansky** [comment]
@cescoffier anything else you want to improve here? thanks.

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
LGTM! That's great (was looking for this tool yesterday while debugging some weird resolution)

**Pull request merged**